packages/babylon/src/tokenizer/context.js

```diff
@@ -103,8 +103,12 @@ tt.incDec.updateContext = function() {
   // tokExprAllowed stays unchanged
 };
 
-tt._function.updateContext = function() {
-  if (this.curContext() !== types.braceStatement) {
+tt._function.updateContext = function(prevType) {
+  if (
+    this.curContext() !== types.braceStatement ||
+    prevType.isAssign ||
```

Comment:
You need to use `this.state.exprAllowed`, because function expression could be introduced by something different from `=`.
e.g.
```js
{
  !function fn() {} /42/i;
}
```

Since if `prevType === tt._return` then `exprAllowed` is true, you could then collapse the two conditions into something like `(this.state.exprAllowed && !this.braceIsBlock)`